cmd/skopeo/login.go

```diff
@@ -0,0 +1,76 @@
+package main
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/containers/common/pkg/auth"
+	"github.com/containers/image/v5/pkg/docker/config"
+	"github.com/containers/image/v5/types"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+type loginOptions struct {
+	global    *globalOptions
+	loginOpts *auth.LoginOptions
+	getLogin  optionalBool
+	tlsVerify optionalBool
+}
+
+func loginCmd(global *globalOptions) *cobra.Command {
+	opts := loginOptions{
+		global:    global,
+		loginOpts: &auth.LoginOptions{},
+	}
+	cmd := &cobra.Command{
+		Use:     "login",
+		Short:   "Login to a container registry",
+		Long:    "Login to a container registry on a specified server.",
+		RunE:    commandAction(opts.run),
+		Example: `skopeo login quay.io`,
+	}
+	usageHelper(cmd)
+	flags := cmd.Flags()
+	optionalBoolFlag(flags, newOptionalBoolValue(&opts.getLogin), "get-login", "Return the current login user for the registry")
+	optionalBoolFlag(flags, newOptionalBoolValue(&opts.tlsVerify), "tls-verify", "Return the current login user for the registry")
+	flags.AddFlagSet(auth.GetLoginFlags(opts.loginOpts))
+	return cmd
+}
+
+func (opts *loginOptions) run(args []string, stdout io.Writer) error {
+	if len(args) > 1 {
+		return errors.Errorf("too many arguments, login takes only 1 argument")
+	}
+	if len(args) == 0 {
+		return errors.Errorf("please specify a registry to login to")
+	}
+	ctx, cancel := opts.global.commandTimeoutContext()
+	defer cancel()
+	sys, err := opts.global.newSystemContext()
+	if err != nil {
+		return err
+	}
+	sys.DockerCertPath = opts.loginOpts.CertDir
+	sys.AuthFilePath = opts.loginOpts.Authfile
+	if opts.tlsVerify.present {
+		sys.DockerDaemonInsecureSkipTLSVerify = !opts.tlsVerify.value
+		sys.DockerInsecureSkipTLSVerify = types.NewOptionalBool(!opts.tlsVerify.value)
+	}
+	server := auth.GetRegistryName(args[0])
```

Comment:
Shouldn’t the following ~15 lines be a part of the shared code as well?

Reply:
podman/buildah have their own parsing CLI functions `SystemContextFromOptions` till use `c.Flag("tls-verify").changed`
These code can be shared if fix of `optionalBoolFlag` in an API and they use optionalBool?

Reply:
Fair enough about the CLI parsing distinction (although it would be good to share that as well, if possible), but I was talking about this line and the following ones. Nothing in the code using `GetRegistryName`/`GetCredentials`/`IdentityToken` is directly CLI-dependent AFAICS.